### ISSUE TITLE
[MGPU] python-gpu/test_large_sizes.py failed.

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#3794]
During test, `weight_` and `labels_` inside `Metainfo` are resharded with a different and non-empty distribution, causing the CHECK inside `reshard` to fail.

The cause is pretty simple, the test first run on single GPU:
https://github.com/dmlc/xgboost/blob/516457fadcd315554e052b24307392b6aad30e49/tests/python-gpu/test_large_sizes.py#L100

Then later run on multi-gpu:
https://github.com/dmlc/xgboost/blob/516457fadcd315554e052b24307392b6aad30e49/tests/python-gpu/test_large_sizes.py#L106

While running multi-gpu test, the distribution of `weigth_` and `label_` remains unchanged.  Hence, to make the test pass we simply have to remove one of these two tests.

This issue is a reminder for me to sort out a better approach for multi-gpu.

**Comment by hcho3** [comment]
@trivialfis Does #3738 solve this issue?

**Comment by trivialfis** [comment]
No, this one needs be solved by handling changed parameter.

**Comment by hcho3** [comment]
@trivialfis I'm confused, don't lines 100 and 106 create Booster object from scratch? Which objects are being shared?

**Comment by hcho3** [comment]
Also, should I add this as a blocking issue?

**Comment by trivialfis** [comment]
@hcho3 The weight and label from DMatrix are not resharded.  To fix this we need to come up with a solution handling  changing parameters, in this case, the number of GPUs . I think we can solve it along with the callback issue, so a known bug describing users currently can't change the number of GPUs in between two training/predict sessions might be more appropriate than a quick fix (reshard everything).

**Comment by hcho3** [comment]
So add an item to "known issue" for the next release, then?

**Comment by trivialfis** [comment]
Yes, please.

**Comment by hcho3** [comment]
@trivialfis What would be the work-around for this issue? Should I delete the DMatrix object and re-load?

**Comment by trivialfis** [comment]
Should work, but let me test it.

**Comment by trivialfis** [comment]
@hcho3 Tested. 
Inserting:
```python
            ag_dtrain = xgb.DMatrix(X, y, nthread=40)
```
In between two training sessions in `/test_large_sizes.py` will let the test pass.  The above line essentially recreated DMatrix.  I will focus on #3825 and #3795 next.  After that we can try to find out a solution to these problems.  Feel free to let me know if you have any suggestion. :)

**Comment by hcho3** [comment]
Got it. The release note for 0.81 will contain a short description of the work-around. Thanks!

**Comment by pseudotensor** [comment]
How does this work-around apply to us using the sklearn interface?

I mention because h2o4gpu testing hasn't failed with xgboost errors in a while, but after recent updates we get the distribution is empty issue for 1 test.
